cpp/src/arrow/io/readahead.cc

```diff
@@ -0,0 +1,222 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include "arrow/io/readahead.h"
+#include "arrow/buffer.h"
+#include "arrow/status.h"
+#include "arrow/util/logging.h"
+
+#include <condition_variable>
+#include <cstring>
+#include <deque>
+#include <memory>
+#include <mutex>
+#include <string>
+#include <thread>
+#include <utility>
+
+namespace arrow {
+namespace io {
+namespace internal {
+
+// ----------------------------------------------------------------------
+// ReadaheadSpooler implementation
+
+class ReadaheadSpooler::Impl {
+ public:
+  Impl(MemoryPool* pool, std::shared_ptr<InputStream> raw, int64_t read_size,
+       int32_t readahead_queue_size, int64_t left_padding, int64_t right_padding)
+      : pool_(pool),
+        raw_(raw),
+        read_size_(read_size),
+        readahead_queue_size_(readahead_queue_size),
+        left_padding_(left_padding),
+        right_padding_(right_padding) {
+    DCHECK_NE(raw, nullptr);
+    DCHECK_GT(read_size, 0);
+    DCHECK_GT(readahead_queue_size, 0);
+    io_worker_ = std::thread([&]() { WorkerLoop(); });
+  }
+
+  ~Impl() { ARROW_UNUSED(Close()); }
+
+  Status Close() {
+    std::unique_lock<std::mutex> lock(mutex_);
+    please_close_ = true;
+    io_wakeup_.notify_one();
+    // Wait for IO thread to finish
+    if (io_worker_.joinable()) {
+      lock.unlock();
+      io_worker_.join();
+      lock.lock();
+    }
+    return raw_->Close();
+  }
+
+  Status Read(ReadaheadBuffer* out) {
+    std::unique_lock<std::mutex> lock(mutex_);
+    while (true) {
+      // Drain queue before querying other flags
+      if (buffer_queue_.size() > 0) {
+        *out = std::move(buffer_queue_.front());
+        DCHECK_NE(out->buffer, nullptr);
+        buffer_queue_.pop_front();
+        // Need to fill up queue again
+        io_wakeup_.notify_one();
+        return Status::OK();
+      }
+      if (!read_status_.ok()) {
+        // Got a read error, bail out
+        return read_status_;
+      }
+      if (eof_) {
+        // XXX maybe we need a EOF status code
```

Comment:
Maybe, can open an issue to consider in more detail. I think it's OK to return null for now